src/aws/flb_aws_imds.c

```diff
@@ -0,0 +1,340 @@
+    ctx->ec2_imds_client = ec2_imds_client;
+    return ctx;
+}
+
+/* Destroy IMDS context */
+void flb_aws_imds_destroy(struct flb_aws_imds *ctx) {
+    if (ctx->imds_v2_token) {
+        flb_sds_destroy(ctx->imds_v2_token);
+    }
+
+    flb_free(ctx);
+}
+
+/* Get IMDS metadata */
+int flb_aws_imds_request(struct flb_aws_imds *ctx, char *metadata_path,
+                        flb_sds_t *metadata, size_t *metadata_len)
+{
+    return flb_aws_imds_request_by_key(ctx, metadata_path, metadata,
+                               metadata_len, NULL);
+}
+
+/* Get IMDS metadata by key */
+int flb_aws_imds_request_by_key(struct flb_aws_imds *ctx, char *metadata_path,
+                               flb_sds_t *metadata, size_t *metadata_len,
+                               char *key)
+{
+    int ret;
+    flb_sds_t tmp;
+    
+    struct flb_http_client *c = NULL;
+
+    struct flb_aws_client *ec2_imds_client = ctx->ec2_imds_client;
+    struct flb_aws_header token_header = imds_v2_token_token_header_template;
+
+    /* Get IMDS version */
+    int imds_version = get_imds_version(ctx);
+
+    /* Abort on version detection failure */
+    if (imds_version == FLB_AWS_IMDS_VERSION_EVALUATE) {
+        // Exit gracefully allowing for retrys
+        flb_warn("[imds] unable to evaluate IMDS version");
+        return -1;
+    }
+
+    if (imds_version == FLB_AWS_IMDS_VERSION_2) {
+        token_header.val = ctx->imds_v2_token;
+        token_header.val_len = ctx->imds_v2_token_len;
+        flb_debug("[imds] using IMDSv2");
+    } else {
+        flb_debug("[imds] using IMDSv1");
+    }
+
+    c = ec2_imds_client->client_vtable->request(ec2_imds_client, FLB_HTTP_GET,
+                                       metadata_path, NULL, 0,
+                                       &token_header,
+                                       (imds_version == FLB_AWS_IMDS_VERSION_1) ? 0 : 1);
+    if (!c) {
+        return -1;
+    }
+
+    /* Detect invalid token */
+    if (imds_version == FLB_AWS_IMDS_VERSION_2 && c->resp.status == 401) {
+        /* TODO: Detect IMDS version requirement change */
+        /* Currently unable to test the following */
+        /* if (imds_version == FLB_AWS_IMDS_VERSION_1) {
+            ctx->imds_version = FLB_AWS_IMDS_VERSION_2;
+        } */
+
+        /* Refresh token and retry request */
+        flb_http_client_destroy(c);
+        ret = refresh_imds_v2_token(ctx);
+        if (ret < 0) {
+            flb_debug("[imds] failed to refresh IMDSv2 token");
+            return -1;
+        }
+        token_header.val = ctx->imds_v2_token;
+        token_header.val_len = ctx->imds_v2_token_len;
+        flb_debug("[imds] refreshed IMDSv2 token");
+        c = ec2_imds_client->client_vtable->request(ec2_imds_client, FLB_HTTP_GET,
+                                       metadata_path, NULL, 0,
+                                       &token_header, 1);
+    }
+
+    if (c->resp.status != 200) {
+        if (c->resp.payload_size > 0) {
```

Comment:
More debug information is good but should we add more checks just for debug log? Just want to discuss this a little bit,